04-store-api/starter/controllers/products.js

```diff
@@ -0,0 +1,127 @@
+  //we will only show the result if featured (or/and others) is/are passed as a query
+  const { featured, company, name, sort, fields, numericFilters } = req.query;
+  const queryObject = {};
+
+  if (featured) {
+    //here we set up a new property to queryObject called .featured and if the value that user pass for featured in url =true we set it to true and only show results with featured:true, if not - we set it  to 'false' and only show results with festured:false
+    queryObject.featured = featured === 'true' ? true : false;
+  }
+
+  //it will look for the value that user provided in the request, and it will look only amoung previous results (featured:true or featured: false) .
+  //for example, if the query is featured=false and company=ikea there will be 6 results (hits)
+  if (company) {
+    queryObject.company = company;
+  }
+
+  //if the name exists we´ll use regex
+  if (name) {
+    queryObject.name = { $regex: name, $options: 'i' };
+    // or queryObject.name=name means if the name exists i ´ll go with my queryObject, and name property will be = to a name that user provides
+  }
+
+  if (numericFilters) {
+    const operatorMap = {
+      '>': '$gt',
+      '>=': '$gte',
+      '=': '$eq',
+      '<': '$lt',
+      '<=': '$lte',
+    };
+
+    //QUESTION 1 TO MY MENTOR: what does this line 55 means? where did it came from?
+    const regEx = /\b(<|>|>=|=|<|<=)\b/g;
+    //if user passes numericFilters (initially it is a string), then i need to replace it with the value that mongoose understand
+    //so i create a fn filters and pass the regular expression (1st parameter(regEx)) (2nd parameter- callback function)
+    
+    // in operatorMap[match] we get access to the value that user provides(match), it could be > , =, etc. and then callback translate in to the mongoose language($...)
+    //so now we can get the result
+    //QUESTION 2 TO MY MENTOR:`-${operatorMap[match]}-` what does this symbols (-)do in this case ?
+    let filters = numericFilters.replace(
+      regEx,
+      (match) => `-${operatorMap[match]}-`
+    );
+    //we can only use numericFilters with numeric values(price, rating)
+    const options = ['price', 'rating'];
+
+    //split method splits string into an array, in this case with ,
+    //then we iterate the array (using .forEach methos) that we got after applying split method
+    filters = filters.split(',').forEach((item) => {
+      //every item of this array will consist of field, operator, value
+      //QUESTION 3 TO MY MENTOR: item.split('-') what does this symbols (-)do in this case ?
+      const [field, operator, value] = item.split('-');
+
+      //if user asks for fields price or rating
+      if (options.includes(field)) {
+        // we will dynamically set up properties on queryObject in mongooose format, we convert string value to Number(). 
+        queryObject[field] = { [operator]: Number(value) };
+      }
+    });
```

Comment:
If we stuck a `console.log(filters)` right _**before**_ this section, we'd see that the `replace` method from earlier has transformed our `numericFilters` from something like `'price>30,rating=4.5'` to `'price-$gt-30,rating-$eq-4.5'`.

Then we split on the comma character (`,`) resulting in a list of strings for each field we want to filter by:

```javascript
const splitFilters = filters.split(',')
console.log(splitFilters);
// ['price-$gt-30', 'rating-$eq-4.5']
```

Then we loop through each value in that list using the [Array.prototype.forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) method. 

On line 74 we take the item we're iterating over, and split it on the dash character (`-`). So for example `'price-$gt-30'` would become a list with three items: `["price", "$gt", "30"]`. We use [destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) to store those results in the variables `field`, `operator`, and `value`.

So in the first loop, `field` = `"price"`, `operator` = `"$gt"`, and `value` = `"30"`. And we add those to our queryObject on line 79. Then on the second loop, `field` = `"rating"`, `operator` = `"$eq"` , and `value` = `"4.5"`

---

Alternatively, if on line 62 we had chosen to use a different character to separate the field from the operator, like for example a semicolon:

```
let filters = numericFilters.replace(
      regEx,
      (match) => `;${operatorMap[match]};`
    );
```

Then we'd need to come and update line 74 to split `'price;$gt;30'` using that character instead (`const [field, operator, value] = item.split(';');`)

And the separator doesn't have to be just one character, we could've even done something like:

```
let filters = numericFilters.replace(
      regEx,
      (match) => `MYSEPARATOR${operatorMap[match]}MYSEPARATOR`
    );
```

Which would give us something like `'priceMYSEPARATOR$gtMYSEPARATOR30'`, which we could split with `const [field, operator, value] = item.split('MYSEPARATOR');`